Sources/Algorithms/Intersperse.swift

```diff
@@ -0,0 +1,149 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// A sequence that presents the elements of a base sequence of elements
+/// with a separator between each of those elements.
+public struct Intersperse<Base: Sequence> {
+  let base: Base
+  let separator: Base.Element
+}
+
+extension Intersperse: Sequence {
+  /// The iterator for an `Intersperse` sequence.
+  public struct Iterator: IteratorProtocol {
+    var iterator: Base.Iterator
+    let separator: Base.Element
+    var state = State.start
+
+    enum State {
+      case start
+      case element(Base.Element)
+      case separator
+    }
+
+    public mutating func next() -> Base.Element? {
+      // After the start, the state flips between element and separator. Before
+      // returning a separator, a check is made for the next element as a
+      // separator is only returned between two elements. The next element is
+      // stored to allow it to be returned in the next iteration.
+      switch state {
+      case .start:
+        state = .separator
+        return iterator.next()
+      case .separator:
+        guard let next = iterator.next() else { return nil }
+        state = .element(next)
+        return separator
+      case .element(let element):
+        state = .separator
+        return element
+      }
+    }
+  }
+
+  public func makeIterator() -> Intersperse<Base>.Iterator {
+    Iterator(iterator: base.makeIterator(), separator: separator)
+  }
+}
+
+extension Intersperse: Collection where Base: Collection {
+  public struct Index: Comparable {
+    enum Kind {
+      case element
+      case separator
+    }
+    let index: Base.Index
+    let kind: Kind
+
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+      if lhs.index < rhs.index { return true }
+      if lhs.index > rhs.index { return false }
+      if lhs.kind == .element, rhs.kind == .separator { return true }
+      return false
+    }
```

Comment:
This looks fine, but if you conform `Kind` to `Comparable` (which will just correctly use the case order I think?) then this becomes as simple as `(lhs.index, lhs.kind) < (rhs.index, rhs.kind)`